### DISCUSSION TITLE
TcpStream::connect_std equivalent in tokio 1.0.0?

### DISCUSSION THREAD

**Discussion by jedisct1** [tokio-rs/tokio#3326] [Q&A] [answered]
`TcpStream::connect_std()` is gone.

`TcpStream::from_std()` exists and includes an example that does a blocking `connect()` and the sets the non-blocking flag and creates a Tokio `TcpStream` from here.

Is there a way to avoid the blocking `connect()`?

**Comment by carllerche** [comment, marked as answer]
[`TcpSocket::connect`](https://docs.rs/tokio/1.0.0/tokio/net/struct.TcpSocket.html#method.connect) lets you configure the socket before connecting. If it is missing a `std` conversion, we can add that.

**Comment by jedisct1** [comment]
What I'm missing is the ability to bind an IP address that is not yet attached to an interface (`IP_BINDANY`/`IPV6_BINDANY`/`IP_FREEBIND`/`SO_BINDANY`).

**Reply by carllerche** [reply]
What sequence of syscalls do you need? `TcpSocket` assumes very little so it should be possible to use.

**Reply by jedisct1** [reply]
Before `bind()`:

- `(IPPROTO_IP, IP_BINDANY)` or `(IPPROTO_IPV6, IPV6_BINDANY)` on FreeBSD
- `(SOL_SOCKET, SO_BINDANY)` on OpenBSD
- `(IPPROTO_IP, IP_FREEBIND)` on Linux

**Reply by carllerche** [reply]
You would:
* create a `tokio::net::TcpSocket`, this just creates the socket (not bound).
* You set the socket option as needed (accessing the raw FD). 
* Optional call `TcpSocket::bind()`
* Call `TcpSocket::connect()`
